File: ios/ExpoSalesforceMIAWModule.swift
import ExpoModulesCore
import SMIClientCore
import SMIClientUI
import UIKit

public class SalesForceMIAWModule: Module {
  private var conversationId: UUID?
  private var uiConfiguration: UIConfiguration?
  
  // Armazenamento para campos de pré-chat
  private var preChatData: [String: String] = [:]
  private var hiddenPreChatData: [String: String] = [:]

  public func definition() -> ModuleDefinition {
    Name("ExpoSalesForceMIAW")

    OnCreate {
      self.conversationId = nil
      self.uiConfiguration = nil
      self.preChatData = [:]
      self.hiddenPreChatData = [:]
    }
    
    Function("configure") { (config: [String: Any]) -> Bool in
      guard let urlString = config["url"] as? String,
            let orgId = config["orgId"] as? String,
            let developerName = config["developerName"] as? String else {
        return false
      }
      print("configure config:", config)
      
      let convId: UUID
      if let conversationIdString = config["conversationId"] as? String,
         let uuid = UUID(uuidString: conversationIdString) {
        convId = uuid
      } else {
        convId = self.getOrCreateConversationId()
      }
      self.conversationId = convId
        
      self.uiConfiguration = UIConfiguration(
        serviceAPI: URL(string: urlString)!,
        organizationId: orgId,
        developerName: developerName,
        conversationId: convId
      )
      
      // Processar campos de pré-chat se fornecidos no configure
      if let preChatFields = config["preChatFields"] as? [String: String] {
        print("configure preChatFields: ", preChatFields)
        self.preChatData = preChatFields
      }
      
      if let hiddenFields = config["hiddenPreChatFields"] as? [String: String] {
        print("configure hiddenFields: ", hiddenFields)
        self.hiddenPreChatData = hiddenFields
      }
      
      return true
    }
    
    Function("setPreChatFields") { (fields: [String: String]) -> Bool in
      self.preChatData = fields
      return true
    }
    
    Function("setHiddenPreChatFields") { (fields: [String: String]) -> Bool in
      self.hiddenPreChatData = fields
      return true
    }
    
    AsyncFunction("openChat") { (promise: Promise) in
      DispatchQueue.main.async {
        guard let config = self.uiConfiguration else {
          promise.reject("ERR_NOT_CONFIGURED", "SDK not configured. Call configure() first.")
          return
        }
        
        guard let rootViewController = self.getRootViewController() else {
          promise.reject("ERR_NO_ROOT_VC", "Could not find root view controller.")
          return
        }
        
        // ============================================
        // IMPLEMENTAÇÃO CORRETA BASEADA NO SDK
        // ============================================
        
        // O SDK usa o Interface com inicializador que aceita preChatFieldValueProvider
        // Passar closures diretamente para o construtor
        
        let chatView: ModalInterfaceViewController
        
        // Se temos campos de pré-chat visíveis OU ocultos, precisamos configurar
        if !self.preChatData.isEmpty || !self.hiddenPreChatData.isEmpty {
          
          // Criar CoreClient para configurar delegates
          let coreClient = CoreFactory.create(withConfig: config)
          
          // 1. Configurar HiddenPreChatDelegate se necessário
          if !self.hiddenPreChatData.isEmpty {
            let hiddenDelegate = HiddenPreChatDelegateImpl(hiddenData: self.hiddenPreChatData)
            coreClient.preChatDelegate = hiddenDelegate
            print("✅ HiddenPreChatDelegate configurado com \(self.hiddenPreChatData.count) campos")
          }
          
          // 2. Criar o chat view com provider de campos visíveis
          if !self.preChatData.isEmpty {
            // O provider é uma closure que modifica os campos
            chatView = ModalInterfaceViewController(
              config,
              preChatFieldValueProvider: { [weak self] preChatFields in
                guard let self = self else { return preChatFields }
                return try await self.modifyPreChatFields(preChatFields)
              }
            )
            print("✅ PreChatFieldValueProvider configurado com \(self.preChatData.count) campos")
          } else {
            chatView = ModalInterfaceViewController(config)
          }
        } else {
          // Sem campos de pré-chat
          chatView = ModalInterfaceViewController(config)
        }
        
        // Apresentar o chat
        rootViewController.present(chatView, animated: true) {
          promise.resolve(true)
        }
      }
    }
    
    AsyncFunction("closeChat") { (promise: Promise) in
      DispatchQueue.main.async {
        guard let rootViewController = self.getRootViewController() else {
          promise.reject("ERR_NO_ROOT_VC", "Could not find root view controller.")
          return
        }
        
        rootViewController.dismiss(animated: true) {
          promise.resolve(true)
        }
      }
    }
    
    Function("getConversationId") { () -> String? in
      return self.conversationId?.uuidString
    }
    
    Function("setConversationId") { (newIdString: String) -> Bool in
      guard let newId = UUID(uuidString: newIdString) else {
        return false
      }
      self.conversationId = newId
      
      if let config = self.uiConfiguration {
        self.uiConfiguration = UIConfiguration(
          serviceAPI: config.serviceAPI,
          organizationId: config.organizationId,
          developerName: config.developerName,
          conversationId: newId
        )
        return true
      }
      return false
    }
    
    Function("clearConversationId") { () -> String in
      let newId = UUID()
      self.conversationId = newId
      UserDefaults.standard.set(newId.uuidString, forKey: "ExpoSalesforceMIAW_ConversationId")
      
      if let config = self.uiConfiguration {
        self.uiConfiguration = UIConfiguration(
          serviceAPI: config.serviceAPI,
          organizationId: config.organizationId,
          developerName: config.developerName,
          conversationId: newId
        )
      }
      return newId.uuidString
    }
    
    Function("registerPushToken") { (token: String) -> Bool in
      // Implementação de registro de token se necessário
      return true
    }
    
    Events("onChatOpened", "onChatClosed", "onMessageReceived", "onError")
  }

  // MARK: - Helper Methods

  private func getOrCreateConversationId() -> UUID {
    let key = "ExpoSalesforceMIAW_ConversationId"
    if let existingIdString = UserDefaults.standard.string(forKey: key),
       let existingUUID = UUID(uuidString: existingIdString) {
      return existingUUID
    }
    let newUUID = UUID()
    UserDefaults.standard.set(newUUID.uuidString, forKey: key)
    return newUUID
  }

  private func getRootViewController() -> UIViewController? {
    guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
          let window = windowScene.windows.first,
          let rootViewController = window.rootViewController else {
      return nil
    }
    
    var topController = rootViewController
    while let presentedViewController = topController.presentedViewController {
      topController = presentedViewController
    }
    return topController
  }
  
  // ============================================
  // MODIFICADOR DE PRÉ-CHAT FIELDS (VISÍVEIS)
  // ============================================
  
  /// Função que modifica os campos de pré-chat com os valores fornecidos
  private func modifyPreChatFields(_ preChatFields: [PreChatField]) async throws -> [PreChatField] {
    print("🔵 modifyPreChatFields chamado com \(preChatFields.count) campos")
    
    var modifiedFields = preChatFields
    
    for (index, field) in modifiedFields.enumerated() {
      let fieldName = field.name
      print("  → Campo: \(fieldName)")
      
      // Se temos um valor para este campo, popula e configura
      if let value = preChatData[fieldName] {
        modifiedFields[index].value = value
        
        // IMPORTANTE: Configurar se o campo é editável ou não
        // true = usuário pode editar o valor preenchido
        // false = campo fica bloqueado (read-only)
        modifiedFields[index].isEditable = false
        
        print("    ✅ Preenchido com: \(value) | isEditable: false")
      }
    }
    
    return modifiedFields
  }
  
  // ============================================
  // DELEGATE PARA HIDDEN PRE-CHAT FIELDS
  // ============================================
  
  /// Delegate para campos OCULTOS (Hidden PreChat Fields)
  /// Estes campos NÃO aparecem na interface, são enviados nos bastidores
  private class HiddenPreChatDelegateImpl: HiddenPreChatDelegate {
    private let hiddenData: [String: String]
    
    init(hiddenData: [String: String]) {
      self.hiddenData = hiddenData
    }
    
    func core(_ core: CoreClient!,
              conversation: Conversation!,
              didRequestPrechatValues hiddenPreChatFields: [HiddenPreChatField]!,
              completionHandler: HiddenPreChatValueCompletion!) {
      
      print("🟣 didRequestPrechatValues (hidden) chamado")
      
      guard var updatedFields = hiddenPreChatFields else {
        print("  ⚠️ Nenhum campo hidden recebido")
        // Retornar array vazio ao invés de dicionário
        completionHandler([])
        return
      }
      
      // Preencher cada campo hidden com o valor correspondente
      for (index, field) in updatedFields.enumerated() {
        let fieldName = field.name
        print("  → Campo hidden: \(fieldName)")
        
        if let value = hiddenData[fieldName] {
          updatedFields[index].value = value
          print("    ✅ Preenchido com: \(value)")
        }
      }
      
      // Retornar os campos atualizados para o SDK
      completionHandler(updatedFields)
    }
  }
}

File: android/src/main/java/expo/salesforcemiaw/ExpoSalesForceMIAWModule.kt
package expo.modules.salesforcemiaw

import android.content.Context
import android.content.SharedPreferences
import androidx.appcompat.app.AppCompatActivity
import com.salesforce.android.smi.core.CoreClient
import com.salesforce.android.smi.core.CoreConfiguration
import com.salesforce.android.smi.core.PreChatField
import com.salesforce.android.smi.core.PreChatValuesProvider
import com.salesforce.android.smi.ui.UIClient
import com.salesforce.android.smi.ui.UIConfiguration
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject
import java.io.InputStream
import java.util.UUID
import android.util.Log

class ExpoSalesForceMIAWModule : Module() {
  private var uiConfiguration: UIConfiguration? = null
  private var coreConfiguration: CoreConfiguration? = null
  private var uiClient: UIClient? = null
  private var coreClient: CoreClient? = null
  private var conversationId: String? = null

  // Armazenamento para campos de pré-chat
  private var preChatData: MutableMap<String, String> = mutableMapOf()
  private var hiddenPreChatData: MutableMap<String, String> = mutableMapOf()

  private val prefs: SharedPreferences by lazy {
    context.getSharedPreferences("ExpoSalesForceMIAW", Context.MODE_PRIVATE)
  }

  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React context is null")

  override fun definition() = ModuleDefinition {
    Name("ExpoSalesForceMIAW")

    // ============================================
    // CONFIGURAR O SDK
    // ============================================
    Function("configure") { config: Map<String, Any?> ->
      try {
        val url = config["url"] as? String ?: return@Function false
        val orgId = config["orgId"] as? String ?: return@Function false
        val developerName = config["developerName"] as? String ?: return@Function false

        Log.d("SalesforceMIAW", "configure called with url=$url, orgId=$orgId, developerName=$developerName")

        // Obter ou criar conversation ID
        val convId = config["conversationId"] as? String ?: getOrCreateConversationId()
        conversationId = convId

        // Processar campos de pré-chat se fornecidos
        (config["preChatFields"] as? Map<String, String>)?.let {
          Log.d("SalesforceMIAW", "preChatFields: $it")
          preChatData.clear()
          preChatData.putAll(it)
        }

        (config["hiddenPreChatFields"] as? Map<String, String>)?.let {
          Log.d("SalesforceMIAW", "hiddenPreChatFields: $it")
          hiddenPreChatData.clear()
          hiddenPreChatData.putAll(it)
        }

        // Criar configuração core
        coreConfiguration = CoreConfiguration.create(
          url = url,
          organizationId = orgId,
          developerName = developerName,
          conversationId = convId
        )

        // Criar configuração UI
        uiConfiguration = UIConfiguration.create(coreConfiguration!!)

        // Criar CoreClient
        coreClient = CoreClient.Factory.create(context, coreConfiguration!!)

        // Registrar provider de hidden fields se necessário
        if (hiddenPreChatData.isNotEmpty()) {
          val hiddenProvider = HiddenPreChatValuesProviderImpl(hiddenPreChatData)
          coreClient?.registerHiddenPreChatValuesProvider(hiddenProvider)
          Log.d("SalesforceMIAW", "✅ HiddenPreChatValuesProvider registrado com ${hiddenPreChatData.size} campos")
        }

        // Criar UIClient
        uiClient = UIClient.createClient(context, uiConfiguration!!)

        // Registrar provider de campos visíveis se necessário
        if (preChatData.isNotEmpty()) {
          uiClient?.preChatFieldValueProvider = { preChatFields ->
            Log.d("SalesforceMIAW", "🔵 preChatFieldValueProvider chamado com ${preChatFields.size} campos")

            preChatFields.map { field ->
              val fieldName = field.name
              Log.d("SalesforceMIAW", "  → Campo: $fieldName")

              // Se temos um valor para este campo, popula e configura
              if (preChatData.containsKey(fieldName)) {
                val value = preChatData[fieldName]!!
                Log.d("SalesforceMIAW", "    ✅ Preenchido com: $value | isEditable: false")

                PreChatField(
                  name = field.name,
                  label = field.label,
                  value = value,
                  isRequired = field.isRequired,
                  isEditable = false, // Campo bloqueado (read-only)
                  isHidden = field.isHidden,
                  maxLength = field.maxLength
                )
              } else {
                field
              }
            }
          }
          Log.d("SalesforceMIAW", "✅ preChatFieldValueProvider configurado com ${preChatData.size} campos")
        }

        true
      } catch (e: Exception) {
        Log.e("SalesforceMIAW", "Erro no configure: ${e.message}", e)
        e.printStackTrace()
        false
      }
    }

    // ============================================
    // DEFINIR CAMPOS DE PRÉ-CHAT VISÍVEIS
    // ============================================
    Function("setPreChatFields") { fields: Map<String, String> ->
      try {
        preChatData.clear()
        preChatData.putAll(fields)
        Log.d("SalesforceMIAW", "setPreChatFields: $fields")
        true
      } catch (e: Exception) {
        Log.e("SalesforceMIAW", "Erro no setPreChatFields: ${e.message}", e)
        false
      }
    }

    // ============================================
    // DEFINIR CAMPOS DE PRÉ-CHAT OCULTOS
    // ============================================
    Function("setHiddenPreChatFields") { fields: Map<String, String> ->
      try {
        hiddenPreChatData.clear()
        hiddenPreChatData.putAll(fields)
        Log.d("SalesforceMIAW", "setHiddenPreChatFields: $fields")

        // Re-registrar o provider se já existe um coreClient
        coreClient?.let { client ->
          val hiddenProvider = HiddenPreChatValuesProviderImpl(hiddenPreChatData)
          client.registerHiddenPreChatValuesProvider(hiddenProvider)
          Log.d("SalesforceMIAW", "✅ HiddenPreChatValuesProvider re-registrado")
        }

        true
      } catch (e: Exception) {
        Log.e("SalesforceMIAW", "Erro no setHiddenPreChatFields: ${e.message}", e)
        false
      }
    }

    // ============================================
    // ABRIR A INTERFACE DE CHAT
    // ============================================
    AsyncFunction("openChat") { promise: Promise ->
      try {
        val client = uiClient ?: run {
          promise.reject("ERR_NOT_CONFIGURED", "SDK not configured. Call configure() first.", null)
          return@AsyncFunction
        }

        val activity = appContext.currentActivity as? AppCompatActivity ?: run {
          promise.reject("ERR_NO_ACTIVITY", "Could not find current activity.", null)
          return@AsyncFunction
        }

        activity.runOnUiThread {
          try {
            Log.d("SalesforceMIAW", "🎨 Abrindo interface do chat...")
            client.openConversation(activity)
            Log.d("SalesforceMIAW", "✅ Chat interface apresentada")
            promise.resolve(true)
          } catch (e: Exception) {
            Log.e("SalesforceMIAW", "Erro ao abrir chat: ${e.message}", e)
            promise.reject("ERR_OPEN_CHAT", "Failed to open chat: ${e.message}", e)
          }
        }
      } catch (e: Exception) {
        Log.e("SalesforceMIAW", "Erro no openChat: ${e.message}", e)
        promise.reject("ERR_OPEN_CHAT", "Failed to open chat: ${e.message}", e)
      }
    }

    // ============================================
    // FECHAR A INTERFACE DE CHAT
    // ============================================
    AsyncFunction("closeChat") { promise: Promise ->
      try {
        val activity = appContext.currentActivity as? AppCompatActivity ?: run {
          promise.reject("ERR_NO_ACTIVITY", "Could not find current activity.", null)
          return@AsyncFunction
        }

        activity.runOnUiThread {
          try {
            // O chat é fechado automaticamente quando o usuário fecha a tela
            Log.d("SalesforceMIAW", "✅ Chat fechado")
            promise.resolve(true)
          } catch (e: Exception) {
            promise.reject("ERR_CLOSE_CHAT", "Failed to close chat: ${e.message}", e)
          }
        }
      } catch (e: Exception) {
        promise.reject("ERR_CLOSE_CHAT", "Failed to close chat: ${e.message}", e)
      }
    }

    // ============================================
    // OBTER O CONVERSATION ID ATUAL
    // ============================================
    Function("getConversationId") {
      conversationId
    }

    // ============================================
    // DEFINIR UM NOVO CONVERSATION ID
    // ============================================
    Function("setConversationId") { newId: String ->
      try {
        conversationId = newId
        prefs.edit().putString("conversationId", newId).apply()

        // Reconfigurar se já existir uma configuração
        coreConfiguration?.let { coreConfig ->
          val newCoreConfig = CoreConfiguration.create(
            url = coreConfig.url,
            organizationId = coreConfig.organizationId,
            developerName = coreConfig.developerName,
            conversationId = newId
          )

          coreConfiguration = newCoreConfig
          uiConfiguration = UIConfiguration.create(newCoreConfig)

          // Recriar clients
          coreClient = CoreClient.Factory.create(context, newCoreConfig)

          // Re-registrar hidden provider se necessário
          if (hiddenPreChatData.isNotEmpty()) {
            val hiddenProvider = HiddenPreChatValuesProviderImpl(hiddenPreChatData)
            coreClient?.registerHiddenPreChatValuesProvider(hiddenProvider)
          }

          uiClient = UIClient.createClient(context, uiConfiguration!!)

          // Re-registrar visible provider se necessário
          if (preChatData.isNotEmpty()) {
            uiClient?.preChatFieldValueProvider = { preChatFields ->
              preChatFields.map { field ->
                if (preChatData.containsKey(field.name)) {
                  PreChatField(
                    name = field.name,
                    label = field.label,
                    value = preChatData[field.name]!!,
                    isRequired = field.isRequired,
                    isEditable = false,
                    isHidden = field.isHidden,
                    maxLength = field.maxLength
                  )
                } else {
                  field
                }
              }
            }
          }

          Log.d("SalesforceMIAW", "ConversationId atualizado para: $newId")
          true
        } ?: false
      } catch (e: Exception) {
        Log.e("SalesforceMIAW", "Erro no setConversationId: ${e.message}", e)
        e.printStackTrace()
        false
      }
    }

    // ============================================
    // LIMPAR O CONVERSATION ID (CRIAR NOVO)
    // ============================================
    Function("clearConversationId") {
      val newId = UUID.randomUUID().toString()
      conversationId = newId
      prefs.edit().putString("conversationId", newId).apply()

      // Reconfigurar se já existir uma configuração
      coreConfiguration?.let { coreConfig ->
        val newCoreConfig = CoreConfiguration.create(
          url = coreConfig.url,
          organizationId = coreConfig.organizationId,
          developerName = coreConfig.developerName,
          conversationId = newId
        )

        coreConfiguration = newCoreConfig
        uiConfiguration = UIConfiguration.create(newCoreConfig)

        // Recriar clients
        coreClient = CoreClient.Factory.create(context, newCoreConfig)

        // Re-registrar hidden provider se necessário
        if (hiddenPreChatData.isNotEmpty()) {
          val hiddenProvider = HiddenPreChatValuesProviderImpl(hiddenPreChatData)
          coreClient?.registerHiddenPreChatValuesProvider(hiddenProvider)
        }

        uiClient = UIClient.createClient(context, uiConfiguration!!)

        // Re-registrar visible provider se necessário
        if (preChatData.isNotEmpty()) {
          uiClient?.preChatFieldValueProvider = { preChatFields ->
            preChatFields.map { field ->
              if (preChatData.containsKey(field.name)) {
                PreChatField(
                  name = field.name,
                  label = field.label,
                  value = preChatData[field.name]!!,
                  isRequired = field.isRequired,
                  isEditable = false,
                  isHidden = field.isHidden,
                  maxLength = field.maxLength
                )
              } else {
                field
              }
            }
          }
        }
      }

      Log.d("SalesforceMIAW", "Nova conversação criada: $newId")
      newId
    }

    // ============================================
    // CONFIGURAR USANDO ARQUIVO config.json
    // ============================================
    Function("configureFromFile") { fileName: String ->
      try {
        val inputStream: InputStream = context.assets.open("$fileName.json")
        val json = inputStream.bufferedReader().use { it.readText() }
        val jsonObject = JSONObject(json)

        val url = jsonObject.getString("url")
        val orgId = jsonObject.getString("orgId")
        val developerName = jsonObject.getString("developerName")

        val convId = getOrCreateConversationId()
        conversationId = convId

        // Criar configuração core
        coreConfiguration = CoreConfiguration.create(
          url = url,
          organizationId = orgId,
          developerName = developerName,
          conversationId = convId
        )

        // Criar configuração UI
        uiConfiguration = UIConfiguration.create(coreConfiguration!!)

        // Criar CoreClient
        coreClient = CoreClient.Factory.create(context, coreConfiguration!!)

        // Criar UIClient
        uiClient = UIClient.createClient(context, uiConfiguration!!)

        Log.d("SalesforceMIAW", "Configurado a partir do arquivo: $fileName.json")
        true
      } catch (e: Exception) {
        Log.e("SalesforceMIAW", "Erro no configureFromFile: ${e.message}", e)
        e.printStackTrace()
        false
      }
    }

    // ============================================
    // REGISTRAR TOKEN DE PUSH NOTIFICATION
    // ============================================
    Function("registerPushToken") { token: String ->
      try {
        // TODO: Implementar registro de push token quando necessário
        Log.d("SalesforceMIAW", "registerPushToken: $token")
        true
      } catch (e: Exception) {
        Log.e("SalesforceMIAW", "Erro no registerPushToken: ${e.message}", e)
        e.printStackTrace()
        false
      }
    }

    // Eventos
    Events("onChatOpened", "onChatClosed", "onMessageReceived", "onError")
  }

  // ============================================
  // HELPER METHODS
  // ============================================

  private fun getOrCreateConversationId(): String {
    val existingId = prefs.getString("conversationId", null)
    if (existingId != null) {
      return existingId
    }

    val newId = UUID.randomUUID().toString()
    prefs.edit().putString("conversationId", newId).apply()
    return newId
  }

  // ============================================
  // PROVIDER PARA HIDDEN PRE-CHAT FIELDS
  // ============================================

  /**
   * Provider para campos OCULTOS (Hidden PreChat Fields)
   * Estes campos NÃO aparecem na interface, são enviados nos bastidores
   */
  private class HiddenPreChatValuesProviderImpl(
    private val hiddenData: Map<String, String>
  ) : PreChatValuesProvider {

    override fun setValues(
      hiddenPreChatFields: List<PreChatField>,
      completionHandler: (List<PreChatField>) -> Unit
    ) {
      Log.d("SalesforceMIAW", "🟣 setValues (hidden) chamado com ${hiddenPreChatFields.size} campos")

      val updatedFields = hiddenPreChatFields.map { field ->
        val fieldName = field.name
        Log.d("SalesforceMIAW", "  → Campo hidden: $fieldName")

        if (hiddenData.containsKey(fieldName)) {
          val value = hiddenData[fieldName]!!
          Log.d("SalesforceMIAW", "    ✅ Preenchido com: $value")

          PreChatField(
            name = field.name,
            label = field.label,
            value = value,
            isRequired = field.isRequired,
            isEditable = field.isEditable,
            isHidden = true,
            maxLength = field.maxLength
          )
        } else {
          field
        }
      }

      // Retornar os campos atualizados para o SDK
      completionHandler(updatedFields)
    }
  }
}